### PR TITLE
Enable docker image cache in the CI workers

### PR DESCRIPTION
**Pull request by v1v**
## What does this pull request do?

Build docker images and cache them in the CI workers using the `.ci/.jenkins_python.yml` as the reference to the docker images to be built

## Why is it important?

Speed up the CI overall execution time

## How to test this

```bash
hub checkout pr 700
sh -x .ci/packer_cache.sh
```
_NOTE_: `/usr/local/bin/bash_standard_lib.sh` is required to exist. That particular requirement is already in place for the CI workers.

## Question
- Using a docker volume when running the docker compose could help with the caching? Or that's something already solved when building the docker image?

### CODE REVIEW

File: tests/scripts/docker/run_tests.sh [hunk 1]
```diff
@@ -44,7 +44,7 @@ fi
 
 # CASS_DRIVER_NO_EXTENSIONS is set so we don't build the Cassandra C-extensions,
 # as this can take several minutes
-docker build --pull --force-rm --build-arg PYTHON_IMAGE=${1/-/:} -t apm-agent-python:${1} . # replace - with : to get the correct docker image
+docker build --force-rm --build-arg PYTHON_IMAGE=${1/-/:} -t apm-agent-python:${1} . # replace - with : to get the correct docker image
```
Comment (kuisathaverat) [inline tests/scripts/docker/run_tests.sh]: The `--force-rm` will invalidate the packer cache

`--force-rm                Always remove intermediate containers`

```suggestion
docker build --build-arg PYTHON_IMAGE=${1/-/:} -t apm-agent-python:${1} . # replace - with : to get the correct docker image
```

**Comment by v1v** [comment]
### How to test this PR

- Create manually the MBP for this repo:
  - Exclude the GH notifications
  - Include only `*700`

<details><summary>Click to view the screenshot</summary>
<p>

image

</p>
</details>

- Build once and stop the build as soon as the pipeline started to execute the checkout
- `Replay` and use the below snippet

<details><summary>Expand to view</summary>
<p>

```
@Library('apm@current') _

import co.elastic.matrix.*
import groovy.transform.Field

@Field def pythonTasksGen

pipeline {
  agent { label 'linux && immutable' }
  environment {
    REPO = 'apm-agent-python'
    BASE_DIR = "src/github.com/elastic/${env.REPO}"
    PIPELINE_LOG_LEVEL='INFO'
    NOTIFY_TO = credentials('notify-to')
    JOB_GCS_BUCKET = credentials('gcs-bucket')
    CODECOV_SECRET = 'secret/apm-team/ci/apm-agent-python-codecov'
    GITHUB_CHECK_ITS_NAME = 'Integration Tests'
    ITS_PIPELINE = 'apm-integration-tests-selector-mbp/master'
    BENCHMARK_SECRET  = 'secret/apm-team/ci/benchmark-cloud'
    OPBEANS_REPO = 'opbeans-python'
    HOME = "${env.WORKSPACE}"
    PATH = "${env.WORKSPACE}/.local/bin:${env.WORKSPACE}/bin:${env.PATH}"
    PIP_CACHE = "${env.WORKSPACE}/.cache"
  }
  options {
    timeout(time: 1, unit: 'HOURS')
    buildDiscarder(logRotator(numToKeepStr: '20', artifactNumToKeepStr: '20', daysToKeepStr: '30'))
    timestamps()
    ansiColor('xterm')
    disableResume()
    durabilityHint('PERFORMANCE_OPTIMIZED')
    rateLimitBuilds(throttle: [count: 60, durationName: 'hour', userBoost: true])
    quietPeriod(10)
  }
  stages {
    stage('Initializing'){
      options { skipDefaultCheckout() }
      stages {
        stage('Checkout') {
          steps {
            deleteDir()
            gitCheckout(basedir: "${BASE_DIR}", githubNotifyFirstTimeContributor: true)
          }
        }
        stage('Build_Packer') {
          steps {
            dir("${BASE_DIR}") {
                sh '.ci/packer_cache.sh'
            }
          }
        }
      }
    }
    /**
    Execute unit tests.
    */
    stage('Test') {
      options { skipDefaultCheckout() }
      steps {
          dir("${BASE_DIR}"){
            script {
              pythonTasksGen = new PythonParallelTaskGenerator(
                xKey: 'PYTHON_VERSION',
                yKey: 'FRAMEWORK',
                xFile: ".ci/.jenkins_python.yml",
                yFile: ".ci/.jenkins_framework.yml",
                exclusionFile: ".ci/.jenkins_exclude.yml",
                tag: "Python",
                name: "Python",
                steps: this
                )
              def mapPatallelTasks = pythonTasksGen.generateParallelTests()
              parallel(mapPatallelTasks)
            }
          }
      }
    }
  }
}

/**
Parallel task generator for the integration tests.
*/
class PythonParallelTaskGenerator extends DefaultParallelTaskGenerator {

  public PythonParallelTaskGenerator(Map params){
    super(params)
  }

  /**
  build a map of closures to be used as parallel steps.
  Make 5 groups per column so it will spin 5 Nodes,
  this makes fewer Docker image Builds.
  */
  protected Map generateParallelSteps(column){
    def parallelStep = [:]
    def groups = [:]
    def index = 1
    column.each{ key, value ->
      def keyGrp = "${this.tag}-${value.x}-${index % 5}"
      if(groups[keyGrp] == null){
        groups[keyGrp] = [:]
        groups[keyGrp].key = value.x
        groups[keyGrp].values = []
      }
      groups[keyGrp].values.add(value.y)
      index++
    }
    groups.each{ key, value ->
      parallelStep[key] = generateStep(value.key, value.values)
    }
    return parallelStep
  }

  /**
  build a clousure that launch and agent and execute the corresponding test script,
  then store the results.
  */
  public Closure generateStep(x, yList){
    return {
      steps.node('linux && immutable'){
        yList.each{ y ->
          def label = "${tag}-${x}-${y}"
          try {
            steps.runScript(label: label, python: x, framework: y)
            saveResult(x, y, 1)
          } catch(e){
            saveResult(x, y, 0)
            steps.error("${label} tests failed : ${e.toString()}\n")
          } finally {
            steps.junit(allowEmptyResults: true,
              keepLongStdio: true,
              testResults: "**/python-agent-junit.xml,**/target/**/TEST-*.xml")
          }
        }
      }
    }
  }
}

def runScript(Map params = [:]){
  def label = params.label
  def python = params.python
  def framework = params.framework
  if (python.endsWith('3.7') || python.endsWith('3.8')) {
      dir("${BASE_DIR}/tests"){
        retry(2){
          sh("docker build --force-rm --build-arg PYTHON_IMAGE=${python.replaceAll('-', ':')} -t apm-agent-python:${python} .")
        }
      }
  }
}

```

</p>
</details>

Expected behavior:

- Docker images have been cached (as there is an additional stage in the pipeline to validate the packer-cache.sh script)

```
 docker images
REPOSITORY                                                      TAG                                        IMAGE ID            CREATED             SIZE
apm-agent-python                                                pypy-3                                     31fc69c89382        10 minutes ago      1GB
apm-agent-python                                                pypy-2                                     218603e88405        11 minutes ago      788MB
apm-agent-python                                                python-3.8                                 0f4c69c55e16        12 minutes ago      975MB
apm-agent-python                                                python-3.7                                 d99f83e14abb        13 minutes ago      959MB
apm-agent-python                                                python-3.6                                 2b8d1c377650        13 minutes ago      954MB
apm-agent-python                                                python-3.5                                 fb3f22f2c26c        14 minutes ago      951MB
apm-agent-python                                                python-2.7                                 660f86889dd8        15 minutes ago      933MB
```

- Docker images are not built again (validating only 3.8 and 3.7 python versions)

```
18:26:38  + docker build --force-rm --build-arg PYTHON_IMAGE=python:3.7 -t apm-agent-python:python-3.7 .
18:26:38  Sending build context to Docker daemon  894.5kB

18:26:38  Step 1/12 : ARG PYTHON_IMAGE
18:26:38  Step 2/12 : FROM ${PYTHON_IMAGE}
18:26:38   ---> de787d71ba24
18:26:38  Step 3/12 : RUN apt-get -qq update && apt-get -qq -y --no-install-recommends install                                                             ca-certificates                                                             curl                                                             netcat                                                             odbc-postgresql                                                             unixodbc-dev                                                             freetds-dev                                                             libmemcached-dev &&    rm -rf /var/lib/apt/lists/*
18:26:38   ---> Using cache
18:26:38   ---> a94659e6a012
18:26:38  Step 4/12 : RUN for server in $(shuf -e ha.pool.sks-keyservers.net                             hkp://p80.pool.sks-keyservers.net:80                             keyserver.ubuntu.com                             hkp://keyserver.ubuntu.com:80                             pgp.mit.edu) ; do gpg --no-tty --keyserver "$server" --recv-keys B42F6819007F00F88E364FD4036A9C25BF357DD4 && s=0 && break || s=$?; done; (exit $s)
18:26:38   ---> Using cache
18:26:38   ---> 9595071520ed
18:26:38  Step 5/12 : RUN curl -o /usr/local/bin/gosu -sSL "https://github.com/tianon/gosu/releases/download/1.10/gosu-$(dpkg --print-architecture)"     && curl -o /usr/local/bin/gosu.asc -sSL "https://github.com/tianon/gosu/releases/download/1.10/gosu-$(dpkg --print-architecture).asc"     && gpg --verify /usr/local/bin/gosu.asc     && rm /usr/local/bin/gosu.asc     && chmod +x /usr/local/bin/gosu
18:26:38   ---> Using cache
18:26:38   ---> e7143c707532
18:26:38  Step 6/12 : COPY entrypoint.sh /usr/local/bin/entrypoint.sh
18:26:38   ---> Using cache
18:26:38   ---> cf858b364c9b
18:26:38  Step 7/12 : COPY requirements/requirements-base.txt /
18:26:38   ---> Using cache
18:26:38   ---> 388e96033399
18:26:38  Step 8/12 : RUN if command -v pypy3; then ln -s $(command -v pypy3) /usr/local/bin/python; elif command -v pypy; then ln -s $(command -v pypy) /usr/local/bin/python; fi
18:26:38   ---> Using cache
18:26:38   ---> 17d78b6c3732
18:26:38  Step 9/12 : RUN python -m pip install --progress-bar off -U pip && pip install --progress-bar off -r /requirements-base.txt
18:26:38   ---> Using cache
18:26:38   ---> 3808d6fde443
18:26:38  Step 10/12 : RUN chmod +x /usr/local/bin/entrypoint.sh
18:26:38   ---> Using cache
18:26:38   ---> 45960529cbb9
18:26:38  Step 11/12 : WORKDIR /app
18:26:38   ---> Using cache
18:26:38   ---> d640fb47904f
18:26:38  Step 12/12 : ENTRYPOINT ["/usr/local/bin/entrypoint.sh"]
18:26:38   ---> Using cache
18:26:38   ---> d99f83e14abb
18:26:38  Successfully built d99f83e14abb
18:26:38  Successfully tagged apm-agent-python:python-3.7
[Pipeline] }
```

File: tests/scripts/docker/run_tests.sh (hunk 1 above)
Comment (v1v) [reply tests/scripts/docker/run_tests.sh]: I did test this particular scenario:
- Given the packer_cache for the existing branch, run the packer_cache.sh script and the `docker build --force-rm --build-arg PYTHON_IMAGE=${1/-/:} -t apm-agent-python:${1} .` and it worked as expected, so it did use the cached docker images.

I also tried to use the cached docker images in the `master` branch by running `docker build --force-rm --build-arg PYTHON_IMAGE=${1/-/:} -t apm-agent-python:${1} .` and it worked as expected, so it did use the cached docker images.

Therefore it's not required the suggestion

Comment (beniwohli) [reply tests/scripts/docker/run_tests.sh]: to be honest, I'm not 100% sure why `--force-rm` is there in the first place. I'd be fine removing it to see if anything breaks.

**Comment by kuisathaverat** [comment]
@v1v we'd move the docker image build to the daily pipeline too here, isn't it?

**Comment by v1v** [comment]
I'll close this PR as there is already a cached docker image process in place:

- See [pipeline](https://apm-ci.elastic.co/job/apm-shared/job/apm-docker-images-pipeline)


image



_NOTE_ the name of the image will be changed in that particular pipeline to match with the ones that used in this project.

**Pull request closed without merging**